src/execution/utils.ts

```diff
@@ -0,0 +1,592 @@
+      hash: Hex
+      sourceChain: number
+      targetChain: number
+    }
+  | {
+      type: 'bundle'
+      id: bigint
+      sourceChain?: number
+      targetChain: number
+    }
+
+interface BundleData {
+  hash: Hex
+  orderPath: OrderPath
+  userOp?: UserOperation
+}
+
+async function prepareTransaction(
+  config: RhinestoneAccountConfig,
+  transaction: Transaction,
+) {
+  const sourceChain =
+    'chain' in transaction ? transaction.chain : transaction.sourceChain
+  const targetChain =
+    'chain' in transaction ? transaction.chain : transaction.targetChain
+  const initialTokenRequests = transaction.tokenRequests
+  const withSession =
+    transaction.signers?.type === 'session' ? transaction.signers.session : null
+  const accountAddress = getAddress(config)
+
+  // Across requires passing some value to repay the solvers
+  const tokenRequests =
+    initialTokenRequests.length === 0
+      ? [
+          {
+            address: zeroAddress,
+            amount: 1n,
+          },
+        ]
+      : initialTokenRequests
+
+  if (withSession) {
+    if (!sourceChain) {
+      throw new Error(
+        `Specifying source chain is required when using smart sessions`,
+      )
+    }
+    // Smart sessions require a UserOp flow
+    return await prepareTransactionAsUserOp(
+      config,
+      sourceChain,
+      targetChain,
+      transaction.calls,
+      transaction.gasLimit,
+      tokenRequests,
+      accountAddress,
+      withSession,
+    )
+  } else {
+    return await prepareTransactionAsIntent(
+      config,
+      sourceChain,
+      targetChain,
+      transaction.calls,
+      transaction.gasLimit,
+      tokenRequests,
+      accountAddress,
+    )
+  }
+}
+
+async function signTransaction(
+  config: RhinestoneAccountConfig,
+  transaction: Transaction,
+  bundleData: BundleData,
+) {
+  const sourceChain =
+    'chain' in transaction ? transaction.chain : transaction.sourceChain
+  const targetChain =
+    'chain' in transaction ? transaction.chain : transaction.targetChain
+  const withSession =
+    transaction.signers?.type === 'session' ? transaction.signers.session : null
+  const accountAddress = getAddress(config)
+
+  if (withSession) {
+    if (!sourceChain) {
+      throw new Error(
+        `Specifying source chain is required when using smart sessions`,
+      )
+    }
+    const userOp = bundleData.userOp
+    if (!userOp) {
+      throw new Error(`User operation is required when using smart sessions`)
+    }
+    // Smart sessions require a UserOp flow
+    return await signUserOp(
+      config,
+      sourceChain,
+      targetChain,
+      accountAddress,
+      withSession,
+      userOp,
+      bundleData.orderPath,
+    )
+  } else {
+    return await signIntent(config, sourceChain, targetChain, bundleData.hash)
+  }
+}
+
+async function submitTransaction(
+  config: RhinestoneAccountConfig,
+  transaction: Transaction,
+  bundleData: BundleData,
+  signature: Hex,
+) {
```

Comment:
Another idea is to bundle all the inputs after each call.

So the flow would be like this:

```ts
const bundleData = await account.prepareTransaction(transaction)
const signedBundleData = await account.prepareTransaction(bundleData)
const transactionResult = await account.submitTransaction(signedBundleData)
```

Reply:
yeah

Reply:
yeah I think this is cleaner